### ISSUE TITLE
externalURL empty configuration error on startup

### ISSUE THREAD

**Issue by beyang** [sourcegraph/sourcegraph#6614]
When a new instance is created, `externalURL` is empty, leading to the following error:

![image](https://user-images.githubusercontent.com/1646931/68903961-78b16c80-06f1-11ea-80ac-6ed3ef653ef6.png)

It also prevents saving the critical configuration until the externalURL is non-empty. This will be confusing for new users.

**Comment by slimsag** [comment]
I definitely agree this isn't the ideal UX here, but this is working as intended currently. This was added in https://github.com/sourcegraph/sourcegraph/pull/5959 to fix this issue: https://github.com/sourcegraph/sourcegraph/issues/3899

To be 100% clear: it is impossible for us to know what the default value should be for a given instance and with the behavior we had previously (where this alert did not show) users were running into major and extremely difficult to debug issues with basic code intel not working: https://github.com/sourcegraph/infrastructure/issues/1514

**Comment by slimsag** [comment]
Also, this is not a release blocker: this same behavior also existed in 3.9.

**Comment by slimsag** [comment]
See also https://sourcegraph.slack.com/archives/C0C324C91/p1572976442082700

**Comment by dadlerj** [comment]
What do you think about putting `externalURL` as a top-level argument in the `docker run` command (and equivalent on K8s) then? If we have to have it ASAP when an instance starts, then perhaps it belongs at that level.

**Comment by slimsag** [comment]
I think that would be a worse experience overall, it would mean that we would provide a command with `-e EXTERNAL_URL=<YOUR EXTERNAL URL>` and if you copy/pasted it, it would fail to run and tell you that you need to insert your external URL.

Nothing prevents the app from starting without the external URL, it's just that full functionality cannot be enabled without it (basic code intel, other extensions, and more). It makes sense to keep it alongside the app configuration.

What is needed here is a better UX / integrating this into our onboarding flow.

**Comment by dadlerj** [comment]
Why wouldn't we just use localhost as the default (versus `<YOUR EXTERNAL URL>`)? My reason for suggestion is that it would be more obvious to people that it needed to be updated (similar to how we have a default port and people figure out how to change that). 

Perhaps if it's not critical, though, it could be an onboarding step for the admin.

Regardless, my vote is to kill the warning if we do nothing else before 3.10. The current experience seems worse than the previous (which I totally agree was also painful, though).

**Comment by slimsag** [comment]
Localhost has the same issue: "I just set up sourcegraph and everything works, but X (code intel, many other things) is very broken, why?" with no indication for us or them what the problem is.

I am curious, have you read through the issue I linked? https://github.com/sourcegraph/infrastructure/issues/1514 after going through this experience, I can unequivocally say my preference is for our onboarding flow to _look worse_ but for site admins to at least know something is misconfigured on their instance.

**Comment by dadlerj** [comment]
Yeah I definitely recall the issue. I also remember fighting removing localhost as the default in a PR (with you and Joe, I believe). 

We just disagree on whether the cure is worse than the disease... Like I said:

> The current experience seems worse than the previous (which I totally agree was also painful, though).

**Comment by dadlerj** [comment]
But to clarify, I'm definitely not the product owner here, and I'm not trying to force the decision, I just want to share my POV as a user.

**Comment by slimsag** [comment]
Makes sense, I just wanted to make sure you didn't miss that context here. I am also not the decision maker here 😛

My only argument here is that for me as a user, I'd rather have the product be ugly/confusing and tell me what I need to do rather than silently break in a way I and support cannot figure out reasonably.

Will re-assign blocker label and leave the decision here up to @christinaforney and @sqs

**Comment by sqs** [comment]
I don't think this issue is a release-blocker. (But this really highlights the urgency of improving our entire configuration and onboarding experience.)

Let's use my wording change in https://github.com/sourcegraph/sourcegraph/pull/6629 and cherry-pick that onto the release branch.

**Comment by unknwon** [comment]
@slimsag I saw you reopened this, would you mind have a comment for the reason so everyone has the context? Maybe the issue title should be renamed/adjusted?

**Comment by slimsag** [comment]
Updated title to reflect current state. Quinn's PR #6629 was a workaround for the larger problem here (that our on-boarding experience configuring externalURL should be nicer overall).

**Comment by slimsag** [comment]
I'm gonna prioritize this in 3.12 or 3.13, not sure which yet. I also just observed a user think they do not need to configure this because they didn't have DNS set up, and then later run into issues caused by it not being set. :/
